Frontend/HsTypeChecker.hs

```diff
@@ -509,42 +510,60 @@ elabHsAlt :: RnMonoTy         {- Type of the scrutinee  -}
+  let (tms, tys)   = unzip binds
+  (rhs_ty, fc_rhs) <- extendCtxTmsM tms tys (elabTerm rhs) -- Type check the right hand side
   storeEqCs        [ res_ty :~: rhs_ty ]         -- All right hand sides should be the same
   return (FcAlt fc_p fc_rhs)
 
+type RnPatBinds = [(RnTmVar, RnPolyTy)]
+
+sortBinds :: RnPatBinds -> RnPatBinds
+sortBinds = sortBy (compare `on` fst)
+
 -- | Elaborate a pattern
-elabHsPat :: RnMonoTy                {- Expected type -}
-          -> RnPat                   {- Pattern       -}
-          -> GenM (TcCtx, FcPat 'Tc) {- Context and elaborated pattern -}
-elabHsPat exp_ty (HsVarPat x)     = do
+elabHsPat :: RnMonoTy                     {- Expected type -}
+          -> RnPat                        {- Pattern       -}
+          -> RnPatBinds                   {- Bindings      -}
+          -> GenM (RnPatBinds, FcPat 'Tc) {- Context and elaborated pattern -}
+elabHsPat exp_ty (HsVarPat x)     binds = do
   a <- TyVar <$> freshRnTyVar KStar              -- Generate fresh type
   storeEqCs [exp_ty :~: a]                       -- Store equivalence
-  ctx <- extendCtxTmM x (monoTyToPolyTy a) (ask) -- Extend context and ask it explicitly
-  return (ctx, FcVarPat $ rnTmVarToFcTmVar x)    -- Return explicit context and elaborated pattern
-elabHsPat exp_ty (HsConPat dc ps) = do
-  (fc_dc, pat_ty, arg_tys) <- elabPatDataCon dc     -- Elaborate data constructor
-  (ctx, fc_ps)             <- elabHsPats arg_tys ps -- Elaborate all patterns together with their matching argument types
-  storeEqCs                [ exp_ty :~: pat_ty ]    -- The expected type must match the pattern type
-  return (ctx, FcConPatNs fc_dc fc_ps)              -- Return total context and elaborate pattern
-elabHsPat _ HsWildPat             = do
+  let binds' = (x, monoTyToPolyTy a) : binds     -- Add new term variable binding to bindings
+  return (binds', FcVarPat $ rnTmVarToFcTmVar x) -- Return explicit context and elaborated pattern
+elabHsPat exp_ty (HsConPat dc ps) binds = do
+  (fc_dc, pat_ty, arg_tys) <- elabPatDataCon dc           -- Elaborate data constructor
+  (binds', fc_ps)          <- elabHsPats arg_tys ps binds -- Elaborate all patterns together with their matching argument types
+  storeEqCs [ exp_ty :~: pat_ty ]                         -- The expected type must match the pattern type
+  return (binds', FcConPatNs fc_dc fc_ps)                 -- Return total context and elaborate pattern
+elabHsPat _ HsWildPat             binds = do
   frsh <- freshFcTmVar
-  ctx  <- ask
-  return (ctx, FcVarPat frsh)
-elabHsPat _ (HsOrPat _ _)         = notImplemented "Or-pattern elaboration not implemented"
+  return (binds, FcVarPat frsh)
+elabHsPat exp_ty (HsOrPat p1 p2)  binds = do
+  (binds1, fcp1) <- elabHsPat exp_ty p1 binds
+  (binds2, fcp2) <- elabHsPat exp_ty p2 binds
+  let zippedBinds = zip (sortBinds binds1) (sortBinds binds2)
+  _ <- mapM_ (\(b1, b2) -> storeEqCs [(toMono $ snd b1) :~: (toMono $ snd b2)]) zippedBinds
```

Comment:
You don't need the `_ <- ` part here; the `mapM_` must return something of type `IO ()` anyway.